cmd/snap-update-ns/change.go

```diff
@@ -70,6 +74,29 @@ func (c *Change) Perform() error {
 		if err != nil {
 			return err
 		}
+		// If the mount point is not present then create a directory in its
+		// place.  This is very naive, doesn't handle read-only file systems
+		// but it is a good starting point for people working with things like
+		// $SNAP_DATA/subdirectory.
+		//
+		// We use lstat to ensure that we don't follow the symlink in case one
+		// was set up by the snap but this is not coded defensively as the code
+		// may execute concurrently with snap application processes.
+		//
+		// To have better defense we may consider freezing the freezer cgroup
+		// belonging to the snap during the execution of snap-update-ns.
```

Comment:
I don't like the sound of this. If the snap can race snap-update-ns, then the snap is able to create arbitrary directories anywhere the snap-update-ns profile allows, which is already pretty wide and by your own admission, will only get wider.

We can do better than this. Recall openat/mkdirat in sc_nonfatal_mkpath() in cmd/libsnap-confine-private/utils.c

Reply:
We can do better wrt mkdir but we cannot ever do better wrt mount itself. I can make the better mkdir for sure but the ultimate race towards mount is unfixable without freezing the snap processes or using some new kernel interface for mounting.

Reply:
I've implemented `SecureMkdirAll` and used it place of the original now.

Reply:
Re mount, yes, we would have to freeze the process since mount() follows symlinks.

Reply:
Ack, I'll do that

Reply:
Done, we are now freezing the group for the update process.